include/internal/catch_run_context.cpp

```diff
@@ -162,14 +162,16 @@ namespace Catch {
         // and should be let to clear themselves out.
         static_cast<void>(m_reporter->assertionEnded(AssertionStats(result, m_messages, m_totals)));
 
+        if (result.getResultType() != ResultWas::Warning)
+            m_messageScopes.clear();
+
```

Comment:
Using `IStreamingReporter::assertionEnded` as-is didn't work. An example assumption in the downstream:
https://github.com/catchorg/Catch2/blob/093b72416de31a804fea15a4055bb44cb523401f/include/reporters/catch_reporter_console.cpp#L364-L366

`UNSCOPED_INFO` messages are cleared in each assertion regardless of the result with its current semantics. I could change above implementation, but it requires shotgun surgery across all reporters inheriting from `StreamingReporterBase`. This also leads to an ambiguity: If `assertionEnded` returns true, will unscoped info be cleared while info will be kept, or will both be cleared?

So, I figured out it's easier to check result type against warning instead of dealing with all these. :)